command/server.go

```diff
@@ -463,13 +463,12 @@ func (c *ServerCommand) Run(args []string) int {
 		return 1
 	}
 
+	// Check if a migration is in progress. If storage isn't immediately
+	// available, start a goroutine to retry the check.
 	migrationStatus, err := CheckMigration(backend)
 	if err != nil {
-		c.UI.Error("Error checking migration status")
-		return 1
-	}
-
-	if migrationStatus != nil {
+		go c.recheckMigration(backend)
```

Comment:
Since this call is non-blocking, wouldn't the command potentially run through the rest of the code before realizing that there is a migration going on and then exit halfway through?

Reply:
I thought migration is an offline operation. If the server can run and potentially write new entries before it's shut down doesn't that make it racy (in the sense that new entries could be written after migration has gone through migrating those paths). The retry interval is relatively low, but isn't there still a chance of writes happening before the server realizes that there's a migration going on and halt with a shutdown?

Reply:
Migration is an offline operation. This check was meant to be a guard against an accidental server start. The version prior to the PR is stricter, with the downside of changing Vault startup behavior for some storage backends. My main concern was that if consul isn't available right at server start, Vault would terminate, which is different than how Vault behaves in production today. I figured it would be better to preserve the current behavior for normal usage, with slightly looser handling in the migration case.

The race condition requires a number of conditions: 1. a migration is occurring  2. operator is choosing to start Vault during a migration 3. storage becomes available only after server start. In this case we'll shut down and alert the operator. No Vault server (source) data is affected, but an inconsistency could get into the migrated data. Perhaps the best approach is to expand the error/shutdown message to state that the destination may be inconsistent and they should rerun the migration, ensuring servers remain offline?

Reply:
Can use a retry logic similar to `ConsulBackend.RunServiceDiscovery`'s approach (i.e. using `time.Timer` or `time.Ticker` and block with a wait group)? That way the server does not terminate if storage is not available yet, but we make sure to perform the migration check before proceeding further.

Reply:
Here's a blocking approach for consideration: https://github.com/hashicorp/vault/commit/e46a9be0f8c4fb88a09ab35a2422e33c4a5f8a20

I do think it's reasonable to block the server until storage is available so long as read errors are emitted. Is there any use case for the server completing startup without storage?